Use CProverString.format as interface to the solver builtin function

romainbrenguier opened the pull request:
The way `String.format` is implemented in the string preprocessing is hacky and buggy. 
By adding a `CProverString.format` function as interface, the preprocessing becomes very simple, the complexity is then transfered to the java models, for which it is much easier to write clean code (than goto-program for java).
This should go together with an update of the models:  https://github.com/diffblue/java-models-library/pull/20

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

smowton reviewed (approved):
Seems sensible

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: e74d75397292e8cb10ae183daede0c0bf103b5f0).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106699730
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 660b193faa6fcc19d7e0491afd47d435638326a2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106716601
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

romainbrenguier committed:
Update java-models-library for diffblue/java-models-library#20

https://github.com/diffblue/java-models-library/pull/20

romainbrenguier committed:
Add preprocessing for CProverString.format

This makes the link between the CProverString.format signature from java
and the ID_cprover_string_format_func from the solver.

romainbrenguier committed:
Refactor add_axioms_for_format_specifier

This makes it accept a function as argument, which gives for possible
types the corresponding value of the argument.
This will allow us to adapt more easily to a change in the way the
arguments of format are given: they are now given as a struct but we
want to have the possibility to give them as a string.

romainbrenguier committed:
Make add_axioms_for_format accept string arguments

Add the possibility to pass argument to the builtin format function as
strings instead of structs with fields.

romainbrenguier committed:
Use non-deprecated function for hex printing

add_axioms_from_int_hex is deprecated, we use
add_axioms_for_string_of_int_with_radix instead.

romainbrenguier committed:
Remove preprocessing for String.format

This should now be in the model, and the implementation of models can
use CProverString.format.

romainbrenguier committed:
Add tests for String.format

This tests the new implementation of String.format using the java models.

romainbrenguier committed:
Remove code that is now unreachable

To use the builtin format function, the java entry point is now
CProverString.format, so we can remove the legacy code which was
handling String.format directly.

romainbrenguier committed:
Update add_axioms_for_format_specifier documentation

Reflects the new way the argument are passed to the format builtin function.

romainbrenguier committed:
Handle the case of null arguments for format

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: fe27d9984e6b5c0ba19d908e3a0e0ff317718957).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106736692
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: c577ab327ff8ed7b9c5a47132f20e790180c120c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106737115
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj commented on src/solvers/strings/string_constraint_generator_format.cpp:
```diff
@@ -267,61 +268,42 @@ get_component_in_struct(const struct_exprt &expr, irep_idt component_name)
+    symbol_generatort &fresh_symbol, const format_specifiert &fs,
+    const std::function<exprt(const irep_idt &)> &get_arg,
+    const typet &index_type, const typet &char_type, array_poolt &array_pool,
+    const messaget &message, const namespacet &ns) {
```
Why all the weird reformatting?

romainbrenguier replied:
my clang-format is broken :sob:

tautschnig replied:
Did #4461 possibly trip it up as you may need to move to a newer version?

romainbrenguier replied:
Oh yes that looks related: 
```
.git-rewrite/t/.clang-format: Invalid argument
YAML:63:21: error: unknown key 'IndentPPDirectives'
IndentPPDirectives: AfterHash
```
I'm using clang-format-6.0

tautschnig replied:
That seems surprising, it's been added in `clang-format` 6 (and `clang-format-6.0` works for me). But you might also try upgrading to `clang-format` 7, which is what we use on Travis.

romainbrenguier replied:
it's not available on ubuntu 16.04

tautschnig replied:
http://apt.llvm.org/ has repositories for Xenial.

allredj commented on src/solvers/strings/string_constraint_generator_format.cpp:
```diff
@@ -299,7 +299,8 @@ add_axioms_for_format_specifier(
   case format_specifiert::STRING:
   {
-    auto string_expr = get_string_expr(array_pool, get_arg("string_expr"));
+    const exprt arg_string = get_arg("string_expr");
```
:pick: typo in commit message ("functtion"). Also there's a missing "of" I think.

allredj commented on jbmc/regression/jbmc-strings/StringFormat/test-hex1.desc:
```diff
@@ -0,0 +1,6 @@
+--function Test.testHex --cp `../../../../scripts/format_classpath.sh . ../../../lib/java-models-library/target/core-models.jar` --property "java::Test.testHex:(I)Ljava/lang/String;.assertion.1"
+^EXIT=10$
+^SIGNAL=0$
+line 7 assertion at file Test.java .*: FAILURE
```
Line numbers don't seem to match. Assertions are at lines 5, 7, 9.

allredj commented on src/solvers/strings/string_constraint_generator_format.cpp:
```diff
@@ -363,21 +365,30 @@ exprt format_arg_from_string(const array_string_exprt &string,
+      equal_exprt{string[1], from_integer('u', string[0].type())},
+      equal_exprt{string[2], from_integer('l', string[0].type())},
+      equal_exprt{string[3], from_integer('l', string[0].type())},
+  };
```
:no_entry_sign: I think there's a missing constraint on the length of the string.

romainbrenguier commented:
@allredj ready for rereview

tautschnig commented on jbmc/src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -1123,57 +1123,37 @@ code_blockt java_string_library_preprocesst::make_assign_function_from_call(
-  {
-    value_type=java_boolean_type();
+  if (type_name == ID_boolean) {
+    value_type = java_boolean_type();
```
Still a lot of reformatting in here?!

romainbrenguier replied:
should be better now

tautschnig replied:
Hmm, then the changes are actually being reverted in a later commit - this still looks very much broken.

tautschnig replied:
It's "Remove code that is now unreachable" that removes the mis-formatted code.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 211fd5ed5a45a7340ccd0dbcf1410dd5d7e59939).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106879072

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 06e7e9f03892c128c7c62e9ecc12f76cd3065070).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106916333
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 7a920a142d350a4387228ae51f0c1ed220fb6c37).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106919141
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: e45a2d0aba7f31b3e383c02679c72aab4eb74a70).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106922534

peterschrammel commented on src/solvers/strings/string_constraint_generator_format.cpp:
```diff
@@ -365,6 +366,55 @@ add_axioms_for_format_specifier(
+    const exprt as_int = format_arg_from_string(string, ID_int);
+    return typecast_exprt{as_int, floatbv_typet{}};
+  }
+  UNHANDLED_CASE;
```
byte, short, double and long are not handled yet?

romainbrenguier replied:
They are all handled by being cast to a long within the model. So ID_int does not need to make a distinction between byte, short, int and long.

tautschnig replied:
Drive-by comment: so what about `double`?

romainbrenguier replied:
It's also cast to a long. ID_float handles both the cases where the formatting string except a float and a double (this was the same before my PR).

romainbrenguier commented:
The java-models-library update is now pointing to master

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 7c34ea4258d32854669aa3dc94084ae37c93fd1e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106943324
